packages/commonwealth/server/routes/etherscanAPI.ts

```diff
@@ -0,0 +1,81 @@
+import { Request, Response, NextFunction } from 'express';
+import axios from 'axios';
+import axiosRetry from 'axios-retry';
+import { networkIdToName } from 'common-common/src/types';
+import { AppError } from '../util/errors';
+import { DB } from '../models';
+import { TypedRequestBody, TypedResponse, success } from '../types';
+import { ETHERSCAN_JS_API_KEY } from '../config';
+
+const fetchEtherscanContract = async (
+  models: DB,
+  req: Request,
+  res: Response
+) => {
+  const { address } = req.body;
+
+  //   First check if the contract already has an abi entry in the database
+  // get Contract Object from Db by address
+  const contract = await models.Contract.findOne({
+    where: { address },
+  });
+  if (contract) {
+    if (contract.abi_id !== null) {
+      throw new AppError('Contract already has an abi entry in the database');
+    } else {
+      // get chain node of contract from DB
+      const chainNode = await models.ChainNode.findOne({
+        where: { id: contract.chain_node_id },
+      });
+      if (!chainNode) {
+        return new AppError('Invalid chain node');
+      }
+      chainNode.eth_chain_id = chainNode.eth_chain_id || 1;
+
+      const network = networkIdToName[chainNode.eth_chain_id];
+
+      const fqdn =
+        network === 'Mainnet' ? 'api' : `api-${network.toLowerCase()}`;
+
+      const url = `https://${fqdn}.etherscan.io/api?module=contract&action=getsourcecode&address=${address}&apikey=${ETHERSCAN_JS_API_KEY}`
```

Comment:
We should check if the ETHERSCAN_JS_API_KEY is set before doing any work -- immediately return if it isn't.